lib/ruby_indexer/lib/ruby_indexer/references_visitor.rb

```diff
@@ -0,0 +1,63 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyIndexer
+  class ReferencesVisitor < YARP::Visitor
+    extend T::Sig
+
+    sig { returns(T::Array[Reference]) }
+    attr_reader :references
+
+    sig { params(index: Index, path: String, fully_qualified_name: String).void }
+    def initialize(index, path, fully_qualified_name)
+      @index = index
+      @path = path
+      @fully_qualified_name = fully_qualified_name
+      @nesting = T.let([], T::Array[String])
```

Comment:
Actually, I tried this refactor and it makes things a little weird for our indexing visitor, since we need to stuff to happen between the push and pop of the stack. I think this might be a bit premature.

Let's wait a bit until we have a clearer picture and then we can refactor with more confidence.

Reply:
I think the superclass can introduce a `visit_class_body` and `visit_module_body` methods for example that are to be redefined in the subclasses. 

So the subclasses do not need to touch `visit_class_node` and get the nesting behavior for free.